invenio_preservation_sync/api.py

```diff
@@ -0,0 +1,163 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2024 CERN.
+#
+# Invenio-Preservation-Sync is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see LICENSE file for more
+# details.
+
+"""Preservation Sync API abstraction for the PreservationInfo entity."""
+
+from dateutil.parser import parse
+from invenio_db import db
+from sqlalchemy.sql import text
+
+from invenio_preservation_sync.errors import PreservationAlreadyReceivedError
+from invenio_preservation_sync.models import PreservationInfo, PreservationStatus
+
+
+class PreservationInfoAPI(object):
+    """
+    API abstraction of a PreservationInfo entity.
+
+    This class provides an abstraction layer for interacting with PreservationInfo objects.
+    It encapsulates the functionality for creating, retrieving, and managing them.
+    """
+
+    query_order = "revision_id desc, archive_timestamp desc, created desc"
+
+    @classmethod
+    def create(
+        cls,
+        record_id,
+        revision_id,
+        status,
+        harvest_timestamp=None,
+        archive_timestamp=None,
+        uri=None,
+        path=None,
+        event=None,
+        description=None,
+    ):
+        """Create a Preservation Information."""
+        status = PreservationInfoAPI.convert_status(status)
+        preservation = PreservationInfo(
+            record_id=record_id,
+            revision_id=revision_id,
+            status=status,
+            harvest_timestamp=harvest_timestamp,
+            archive_timestamp=archive_timestamp,
+            uri=uri,
+            path=path,
+            event=event,
+            description=description,
+        )
+
+        db.session.add(preservation)
+        db.session.commit()
+        return preservation
+
+    @classmethod
+    def get_by_record_id(cls, record_id):
+        """Get Preservation Info by record id."""
+        preservations = PreservationInfo.query.filter_by(record_id=record_id).order_by(
+            text(cls.query_order)
+        )
+        return preservations
+
+    @classmethod
+    def get_latest_by_record_id(cls, record_id):
+        """Get latest Preservation Info for the given record id."""
+        preservation = (
+            PreservationInfo.query.filter_by(record_id=record_id)
+            .order_by(text(cls.query_order))
+            .first()
+        )
+        return preservation
```

Comment:
Are both needed? [first()](https://docs.sqlalchemy.org/en/14/orm/query.html#sqlalchemy.orm.Query.first) just returns the first of the result. What determines the `latest` is the `query_order`.

Suggestion: have one method only, move the `query_order` inside

Two options: 

1. Make the query lazy and let the caller be responsible for how to use it:

```python

return (
    PreservationInfo.query.filter_by(record_id = record_id)
        .order_by(
            PreservationInfo.revision_id.desc(),
            PreservationInfo.archive_timestamp.desc(),
            PreservationInfo.created.desc())
)
```

2. Add a parameter to the method and build the query:

```python

    @classmethod
    def get_by_record_id(cls, record_id=None, latest=False):

        preservation = PreservationInfo.query.filter_by(record_id=record_id).order_by(
            PreservationInfo.revision_id.desc(),
            PreservationInfo.archive_timestamp.desc(),
            PreservationInfo.created.desc(),
        )
        if latest:
            return preservation.first()
        return preservation  # or preservation.all() if you want to execute the query immediately

```